### PR TITLE
Issue #4934: Enforced WS in appropriate places for block javadoc tags

### PR DESCRIPTION
**Pull request by ps-sp**
Issue #4934 : Enforced WS in appropriate places for block javadoc tags

### CODE REVIEW

**Comment by ps-sp** [comment]
[diffreport-JavadocParagraph](https://4934-diffreport1.netlify.com/)
[ANTLR diff](https://4934-antlrdiff.netlify.com/)

The diffs are due to wrong javadoc which resulted in parse error and so the diffs should be acceptable

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4 [hunk 1]
```diff
@@ -866,32 +866,33 @@ reference:
...
 parameters: LEFT_BRACE (ARGUMENT | COMMA | WS | NEWLINE | LEADING_ASTERISK)* RIGHT_BRACE;
 
-javadocTag: AUTHOR_LITERAL (WS | NEWLINE)* description?
+javadocTag: AUTHOR_LITERAL (WS | NEWLINE)+ description?
 
-    | DEPRECATED_LITERAL (WS | NEWLINE)* description?
+    | DEPRECATED_LITERAL (WS | NEWLINE)* ((WS | NEWLINE) description)?
 
-      | EXCEPTION_LITERAL (WS | NEWLINE)* CLASS_NAME? (WS | NEWLINE)* description?
+      | EXCEPTION_LITERAL (WS | NEWLINE)+ CLASS_NAME? (WS | NEWLINE)* ((WS | NEWLINE) description)?
```
Comment (Vladlis) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: `(WS | NEWLINE)+` - will this work for `/** @exception*/` ? Please add such inputs in UTs for every tag

Comment (ps-sp) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammars/javadoc/JavadocParser.g4]: @vladis

Updated PR please have a look.

**IMPORTANT:** Such javadocs are wrong and should be corrected in #4942

**Comment by Vladlis** [comment]
@rnveach , @romani , Please review

**Comment by rnveach** [comment]
I don't see any changes in code coverage in pom. Did the percentage stay the same and not increase?

**Comment by ps-sp** [comment]
@rnveach 

Sorry I missed your comment. The coverage had decreased as the grammar was modified. Therefore the coverage was restored instead of getting increased. We should have covered "almost" all new grammar code as all that was done was changing the way `(WS | NEWLINE)` was to be treated and there were already tons of cases where javadoc tags had `(WS | NEWLINE)` before their arguments.

**Comment by romani** [comment]
@ps-sp , please do rebase.

please add test case from issue. Issue test case always have to be in UT cases.

=============


Note:
> `(WS | NEWLINE)* ((WS | NEWLINE) description)?`

should become `(WS | NEWLINE)+ description` in scope of #4942 , I approve the issue.

**Comment by ps-sp** [comment]
@romani 
Rebased.

>Parse error expected from ANTLR.

Done. Added one more UT. UT is in `DetailNodeTreeStringPrinterTest`.

**Comment by Vladlis** [comment]
@ps-sp please add such UTs for every affected tag

**Comment by ps-sp** [comment]
I had expected `JavadocParser` coverage to decrease a bit when upon removing the examples mentioned in the above cases. But apparently it didn't. Also, `JavadocLexer` coverage has increased which is fine.

Please correct me if wrong but I have not been able to find any package for which the coverage decreased in cobertura reports, still the total line and total branch coverage decreased by 1 %. May be some rounding off might be involved here and there. 

Please have a look @Vladlis

**Comment by Vladlis** [comment]
@ps-sp , please enforce `lf` line ending for problematic cases in https://github.com/checkstyle/checkstyle/blob/master/.gitattributes until #4980

**Comment by ps-sp** [comment]
@Vladlis 
>please enforce lf line ending for problematic cases in https://github.com/checkstyle/checkstyle/blob/master/.gitattributes until #4980

Done.

**Comment by Vladlis** [comment]
@romani , all test cases are added, please review

**Pull request merged**